File: my_code/DSA/JS/DS/linked-lists/doubly-linked-list/doublyLinkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  // add a new node at the end of the dll
  push(val) {
    // make the new node
    const newNode = new Node(val);

    if (!this.head) {
      // if empty => make the head and the tail to be the newNode
      this.head = newNode;
      this.tail = newNode;
    } else {
      // make the tail poin into the new node
      this.tail.next = newNode;
      //   set the new node to point back to the tail
      newNode.prev = this.tail;
      // update the tail to be the new node
      this.tail = newNode;
    }
    // always increment the length
    this.length++;
    // return the new list
    return this;
  }

  //remove the last node
  pop() {
    // if empty
    if (!this.head) {
      return undefined;
    }

    // store it to return it after removing
    const poppedNode = this.tail;
    // if just one node => return to the initial point
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    }
    //
    else {
      // set the tail to be the previous node
      this.tail = poppedNode.prev;
      // set the new tail next to point into null
      this.tail.next = null;
      // set the old tail next to point to null
      // (UNLINK)
      poppedNode.prev = null;
    }
    // decrement the length
    this.length--;
    // return the popped node
    return poppedNode;
  }

  // remove the head
  shift() {
    if (!this.head) {
      return undefined;
    }

    // store the head to return it later
    const oldHead = this.head;

    // if just one node => return to the initial point
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      // set the head to be the new one (second node)
      this.head = oldHead.next;
      // set the new head prev to point to None
      this.head.prev = null;
      // set the old head next to point to None
      oldHead.next = null;
    }
    this.length--;
    return oldHead;
  }

  // add a new node as head
  unshift(val) {
    // first create a new node to add it
    const newNode = new Node(val);

    // if empty make the head and tail to be the new node
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      // set the new node next to point into the old head
      newNode.next = this.head;
      // then set the old head prev to point back into the new node
      this.head.prev = newNode;
      // then update the head to be the new node
      this.head = newNode;
    }
    // increment the length
    this.length++;
    // then return the list
    return this;
  }

  // get a node based on specific index
  get(index) {
    // invalid index
    if (index < 0 || index >= this.length) {
      return null;
    }

    let current, counter;
    // if closer to the head
    // start from the head
    if (index <= this.length / 2) {
      current = this.head;
      counter = 0;
      while (counter < index) {
        current = current.next;
        counter++;
      }
      return current;
    }
    // if closer to the tail
    // start from the tail
    else {
      current = this.tail;
      counter = this.length - 1;
      while (current > index) {
        current = current.prev;
        counter--;
      }
      return current;
    }
  }

  // set a node value with a new one
  set(index, val) {
    // use the get method to get the node
    const foundNode = this.get(index);

    // if the node exist
    if (foundNode) {
      // set its value to be the new one
      foundNode.val = val;
      return true;
    } else {
      return false;
    }
  }

  // insert a new node based on index
  insert(index, val) {
    // invalid index
    if (index < 0 || index > this.length) {
      return false;
    }
    // if the first node => unshift
    else if (index === 0) {
      this.unshift(val);
      return true;
    }
    // if the last node length => push
    else if (index === this.length) {
      this.push(val);
      return true;
    } else {
      // create a new node
      const newNode = new Node(val);

      // get the node before index we want add
      const beforeNode = this.get(index - 1);
      // get the node after (more readable)
      const afterNode = beforeNode.next;

      //  set the new node next to point into the after node
      newNode.next = afterNode;
      //  and the after node prev to point back to the new node
      afterNode.prev = newNode;
      //  set the before node next to point into new node
      beforeNode.next = newNode;
      // and the new node prev to point back into the before node
      newNode.prev = beforeNode;

      this.length++;
      return true;
    }
  }

  // remove a node based on specific index
  remove(index) {
    // invalid index here also the length not valid
    if (index < 0 || index >= this.length) {
      return undefined;
    }
    // remove from first
    else if (index === 0) {
      return this.shift();
    }
    // remove from last
    else if (index === this.length - 1) {
      return this.pop();
    }
    // remove from inside the list
    else {
      // get the node will remove
      let foundNode = this.get(index);
      // the one before it
      let beforeNode = foundNode.prev;
      // the one after it
      let afterNode = foundNode.next;

      // then make the before node jump and point on after node
      beforeNode.next = afterNode;
      // then make the after node jump and point back to before node
      afterNode.prev = beforeNode;

      // unlink the removed node
      foundNode.next = null;
      foundNode.prev = null;

      this.length--;
      return foundNode;
    }
  }

  // reverse a doubly linked list
  reverse() {
    let current = this.head;
    while (current) {
      // Swap next and prev for current node
      [current.prev, current.next] = [current.next, current.prev];
      // Move to the next node in original order (which is prev now)
      current = current.prev;
    }
    // Swap head and tail pointers
    [this.head, this.tail] = [this.tail, this.head];
  }

  // print
  print() {
    let res = "";
    let current = this.head;
    while (current) {
      res += current.val + " ";
      current = current.next;
    }
    console.log(res);
  }
}

list = new DoublyLinkedList();
list.push(10).push(20).push(30);
console.log("After Reverse:");
list.print();
list.reverse();
console.log("After Reverse:");
list.print();

File: my_code/DSA/PYTHON/DS/linked lists/doubly-linked-list/doubly_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    # add a new node at the end of the dll
    def push(self, val):
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node

        else:
            # set the tail to point into the new_node
            self.tail.next = new_node
            # and set the new_node to point back to the tail
            new_node.prev = self.tail
            # then update the tail
            self.tail = new_node
        
        self.length += 1
        return self
    
    # remove the last node
    def pop(self):
        # if empty
        if not self.head:
            return None
        
        # store it to return it after removing
        popped_node = self.tail

        # if just one node => return to the initial point
        if self.length == 1:
            self.head = None
            self.tail = None

        else:
            # set the tail to be the previous node
            self.tail = popped_node.prev
            # set the new tail next to point into None
            self.tail.next = None
            # set the old tail next to point to None
            # (Unlink)
            popped_node.prev = None
        # decrement the length
        self.length -= 1
        # return the popped node
        return popped_node

    # remove the head
    def shift(self):
        if not self.head:
            return None
        
        # store the head to return it later
        old_head = self.head

        # if just one node => return to the initial point
        if self.length == 1:
            self.head = None
            self.tail = None
        
        else:
            # set the head to be the new one (second node)
            self.head = old_head.next

            # set the new head prev to point to None
            self.head.prev = None

            # set the old head next to point to None
            old_head.next = None 
        self.length -= 1
        return old_head

    # add a new node as head
    def unshift(self, val):
        # first create a new node to add it
        new_node = Node(val)

        # if empty make the head and tail to be the new node
        if not self.head:
            self.head = new_node
            self.tail = new_node
        

        else:
            # set the new node next to point into the old head
            new_node.next = self.head
            # then set the old head prev to point into the new node
            self.head.prev = new_node
            # then update the head to be the new node
            self.head = new_node
        # increment one to the length
        self.length += 1
        # return the list
        return self

    # get a node based on specific index
    def get(self, index):
        # invalid index
        if index < 0 or index >= self.length:
            return None
        
        # if closer to the head
        # start from the head
        if index <= self.length // 2:
            current = self.head
            counter = 0
            while counter != index:
                current = current.next
                counter += 1
            return current
        # if closer to the tail
        # start from the tail
        else:
            current = self.tail
            counter = self.length - 1
            while counter != index:
                current = current.prev
                counter -= 1
            return current

    # set a node with a new value
    def set(self, index, value):
        # use the get method to get the node
        found_node = self.get(index)

        # if the node exist
        if found_node:
            # set its value to be the new one
            found_node.val = value
            return True
        else:
            return False

    # insert a new node based on index
    def insert(self, index, val):
        if index < 0 or index > self.length:
            return False
        elif index == 0:
            self.unshift(val)
        elif index == self.length:
            self.push(val)
        
        else:
            # create a new node
            new_node = Node(val)
            # get the node before index we want add
            before_node = self.get(index - 1)
            # get the node after (more readable)
            after_node = before_node.next

            # set the new node next to point into the after node
            new_node.next = after_node
            # and the after node prev to point back to the new node
            after_node.prev = new_node
            # set the before node next to point into new node
            before_node.next = new_node
            # and the new node prev to point back into the before node
            new_node.prev = before_node

            self.length += 1
            return True




    def __str__(self):
        res = ''
        current = self.head
        while current:
            res += str(current.val) + ' '
            current = current.next
        return res
    
l = DoublyLinkedList()
l.push(10).push(20).push(30)
print(l)
l.insert(1, 15)
print(l)